pex/requirements.py

```diff
@@ -5,67 +5,586 @@
 
 import os
 import re
+import ssl
+import time
+from collections import namedtuple
+from contextlib import closing, contextmanager
 
+from pex.compatibility import (
+    HTTPError,
+    HTTPSHandler,
+    ProxyHandler,
+    Request,
+    build_opener,
+    to_unicode,
+    urlparse,
+)
+from pex.network_configuration import NetworkConfiguration
+from pex.third_party.packaging.markers import Marker
+from pex.third_party.pkg_resources import Requirement, RequirementParseError
+from pex.typing import TYPE_CHECKING, cast
 
-class UnsupportedLine(Exception):
-    pass
+if TYPE_CHECKING:
+    from typing import BinaryIO, Dict, Iterator, Match, Optional, Text, Tuple, Union, Iterable
+
+
+class LogicalLine(
+    namedtuple("LogicalLine", ["raw_text", "processed_text", "source", "start_line", "end_line"])
+):
+    @property
+    def raw_text(self):
+        # type: () -> str
+        return cast(str, super(LogicalLine, self).raw_text)
+
+    @property
+    def processed_text(self):
+        # type: () -> str
+        return cast(str, super(LogicalLine, self).processed_text)
+
+    @property
+    def source(self):
+        # type: () -> str
+        return cast(str, super(LogicalLine, self).source)
+
+    @property
+    def start_line(self):
+        # type: () -> int
+        return cast(int, super(LogicalLine, self).start_line)
+
+    @property
+    def end_line(self):
+        # type: () -> int
+        return cast(int, super(LogicalLine, self).end_line)
+
+    def render_location(self):
+        # type: () -> str
+        if self.start_line == self.end_line:
+            return "{} line {}".format(self.source, self.start_line)
+        return "{} lines {}-{}".format(self.source, self.start_line, self.end_line)
+
+
+class URLFetcher(object):
+    def __init__(self, network_configuration=None):
+        # type: (Optional[NetworkConfiguration]) -> None
+        network_configuration = network_configuration or NetworkConfiguration.create()
+
+        self._timeout = network_configuration.timeout
+        self._max_retries = network_configuration.retries
+        self._headers = network_configuration.headers_as_dict()
+
+        ssl_context = ssl.create_default_context(cafile=network_configuration.cert)
+        if network_configuration.client_cert:
+            ssl_context.load_cert_chain(network_configuration.client_cert)
+
+        proxies = None  # type: Optional[Dict[str, str]]
+        if network_configuration.proxy:
+            proxies = {
+                protocol: network_configuration.proxy for protocol in ("ftp", "http", "https")
+            }
+
+        self._handlers = (ProxyHandler(proxies), HTTPSHandler(context=ssl_context))
+
+    @contextmanager
+    def get_body_iter(self, url):
```

Comment:
I tested all this out - except client certs - manually using `mitmproxy`. Since `mitmproxy` is implemented in python and does have an API it may be reasonable to circle back and add that into an intergation test.

Reply:
Shall we open an issue for this?